File: examples/protocols-interfaces-traits/stack.rs
trait Stackable<T> {
  fn push(&mut self, item: T) -> ();
  fn pop(&mut self) -> Option<T>;
  fn peak(&self) -> Option<&T>;
}

struct Stack<T> {
  items: Vec<T>,
}

impl<T> Stackable<T> for Stack<T> {
  fn push(&mut self, item: T) {
    self.items.push(item);
  }

  fn pop(&mut self) -> Option<T> {
    return self.items.pop();
  }

  fn peak(&self) -> Option<&T> {
    return self.items.last();
  }
}

impl<T> Stack<T> {
  fn new() -> Stack<T> {
    Stack { items: Vec::new() }
  }
}

fn main() {
  let mut s1 = Stack::new();
  s1.push(1);
  s1.push(2);
  s1.pop();
  print!("{:?}", s1.peak())
}

File: examples/protocols-interfaces-traits/stack.kt
interface Stackable<T> {
  fun push(item: T)
  fun pop(): T?
  fun peak(): T?
}

class Stack<T> : Stackable<T> {
  var items: MutableList<T> = mutableListOf()

  override fun push(item: T) {
    items.add(item)
  }

  override fun pop(): T? {
    return items.removeAt(items.size - 1);
  }

  override fun peak(): T? {
    return items.last();
  }
}

fun main() {
  val s1 = Stack<Int>();
  s1.push(1)
  s1.push(2)
  s1.pop()
  println(s1.peak())
}

